clippy_lints/src/methods/mod.rs

```diff
@@ -3169,3 +3205,26 @@ fn check_pointer_offset(cx: &LateContext<'_, '_>, expr: &hir::Expr<'_>, args: &[
         }
     }
 }
+
+fn lint_filetype_is_file(cx: &LateContext<'_, '_>, expr: &hir::Expr<'_>, args: &[hir::Expr<'_>]) {
+    let ty = cx.tables.expr_ty(&args[0]);
+
+    if !match_type(cx, ty, &paths::FILE_TYPE) {
+        return;
+    }
+
+    if_chain! {
+        if let Some(parent) = get_parent_expr(cx, expr);
+        if let hir::ExprKind::Unary(op, _) = parent.kind;
+        if op == hir::UnNot;
+        then {
+            let lint_msg = "`!FileType::is_file()` does not deny all readable file types";
+            let help_msg = "use `FileType::is_dir()` instead";
+            span_help_and_lint(cx, FILETYPE_IS_FILE, parent.span, lint_msg, help_msg);
+        } else {
+            let lint_msg = "`FileType::is_file()` does not cover all readable file types";
+            let help_msg = "use `!FileType::is_dir()` instead";
+            span_help_and_lint(cx, FILETYPE_IS_FILE, expr.span, lint_msg, help_msg);
```

Comment:
IMO a better lint message would be "
```rust
let lint_msg = format!("`{}FileType::is_file()` only covers regular files", /*either "!" or "", depending on the case*/);
let help_msg = format!("use `{}FileType::is_dir()` instead", /*either "!" or "", depending on the case*/);
```

Reply:
I am not sure it is ok to say `!FileType::is_file()` only covers regular files? Because `!FileType.is_file()` is true only if FileType is a dir. In this case `FileType::is_file()` covers regular files, but `!FileType::is_file()` (with `!`) does not.

Reply:
`!FileType.is_file()` should also be true for symlinks, shouldn't it?

Reply:
Eh, that's not what I meant. I am not an native English speaker, and I am not sure "`!FileType::is_file()` only covers regular files" would be perceived as when FileType is a regular file return true not as when FileType is not a regular file return true.